Series Flush Type

ax3l opened the pull request:
Add a new potential data flush mode to `Series`.

Although we implement a more general model of data passing and accumulating, allowing a "directly synchronous" mode will allow us to target the typical h5py Python user who just quickly wants to explore small data sets without frustration ("why is my data all zeros?").

We definitely have a continued need for our generalized, deferred, asynchronous chunk load/store mechanisms for our applications in parallel I/O and upcoming staging (think: streaming) modes. Just let us enable them explicitly as an "advanced" feature (therefore switch the default to the simple `DIRECT` mode). This is purely a user-interface change in order to target a wider audience.

The new `DIRECT` mode should also help somewhat in debugging.

Stylistic note: I think `AccessType` and `FlushType` might better be called `...Mode`?

@franzpoeschel @C0nsultant what do you think? I hope that description is not too brief, otherwise let's chat :-)
PRs to this branch that implement that mode, also in parts, very welcome.

## To Do

- [x] python bindings
- [ ] bugfix of #490
- Docs
  - [x] update first read / write
  - [x] update store/loadChunk doc strings
  - [x] add changelog
  - [x] breaking: add in upgrade guide
- [x] update examples
- [x] update and add more tests

ax3l committed:
Data Flush Mode

Add a new data flush mode to Series.

Although we implement a more general model of data passing
and accumulating, allowing a "directly synchronous" mode will
allow us to target the typical h5py Python user who just quickly
wants to explore small data sets without frustration.

We definitely have a continued need for our generalized, deferred,
asynchronous chunk load/store mechanisms for our applications in
parallel I/O and upcoming staging (think: streaming) modes. Just
let us enable them explicitly as an "advanced" feature.

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -129,6 +136,14 @@ class RecordComponent : public BaseRecordComponent
+     * Series::flush is called.
+     * @todo implement: if the shared pointers' `use_count` drops to one
+     *       before Series::flush is called, the load will be skipped during
+     *       Series::flush. (check if this works with our `shareRaw()`
```
this will become possible when #470 is implemented

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -137,9 +152,37 @@ class RecordComponent : public BaseRecordComponent
+     * Series::flush is called.
+     * @todo implement: if the shared pointers' `use_count` drops to one
+     *       before Series::flush is called, the store will be skipped during
+     *       Series::flush. (check if this works with our `shareRaw()`
```
this will become possible when #470 is implemented

ax3l commented:
@C0nsultant do you think the right location to store the `flush mode` would be with `AbstractIOHandler`?

ax3l committed:
Tests: FlushType

Try to use deferred mode for most tests, but add some
direct ones as well.

Especially, with fixed defaults for particle position/positionOffset
creation, we should add direct mode tests for CREATE of those as
well.

ax3l commented on test/ParallelIOTest.cpp:
```diff
@@ -86,6 +86,7 @@ TEST_CASE( "hdf5_write_test", "[parallel][hdf5]" )
     uint64_t mpi_size = static_cast<uint64_t>(mpi_s);
     uint64_t mpi_rank = static_cast<uint64_t>(mpi_r);
     Series o = Series("../samples/parallel_write.h5", AccessType::CREATE, MPI_COMM_WORLD);
+    o.setFlush(FlushType::DEFER);  // @todo crashes in DIRECT flush mode
```
see bug #490

ax3l commented on test/ParallelIOTest.cpp:
```diff
@@ -120,6 +123,7 @@ TEST_CASE( "hdf5_write_test_zero_extent", "[parallel][hdf5]" )
     uint64_t size = static_cast<uint64_t>(mpi_s);
     uint64_t rank = static_cast<uint64_t>(mpi_r);
     Series o = Series("../samples/parallel_write_zero_extent.h5", AccessType::CREATE, MPI_COMM_WORLD);
+    o.setFlush(FlushType::DEFER);  // @todo crashes in DIRECT flush mode
```
see bug #490

anokfireball commented:
> @C0nsultant do you think the right location to store the `flush mode` would be with `AbstractIOHandler`?

There (currently) is a 1:1 mapping of `Series` to `IOHandlers`, but in theory this could become n:1 in the future (i.e. multiple versions of a `Series` in the same directory with the same access properties). One could think of scenarios where the different versions have different `flush` strategies, but I would not worry much about that case as there's one too many _if_ s involved for serious dedication.

Strictly speaking separation of concern, the IO backend (and in this case, the handler) is a perfectly suitable place for it. The `flush()` calls in the front- and backend have slightly different intentions. The one you care about (actually swapping bytes between RAM and NVstorage, without traversing and syncing the whole frontend tree) is the one the backend is concerned with.

ax3l commented:
Thanks, sounds good.

I actually pushed a first working draft in this PR, for in-code comments/review :)

ax3l committed:
Python: Flush Type

ax3l committed:
Examples: Flush Type

ax3l committed:
MPI Benchmark: Flush Type

ax3l committed:
Changelog & Upgrade: Flush Type

Document the new Series flush type and how to upgrade from
previous releases, as this is a breaking change.

ax3l committed:
First Read/Write: Flush Type

ax3l commented:
@C0nsultant ready for review :)
Any idea why `direct` writing in `3_write_serial` might be segfaulting? Do we need to flush some path creations? Seems the `Record::flush_impl` path creation is a bit late as well as the parent/writable registration for new record components if they are not scalar.

ax3l committed:
Debug: concrete_h5_file_position

dynamic casts can return nullptrs ;-)

  https://github.com/openPMD/openPMD-api/issues/471

ax3l commented:
@C0nsultant if you have the time, please feel free to add additional commits

anokfireball commented on NEWS.rst:
```diff
@@ -3,6 +3,56 @@
+``.flush()`` Method
+"""""""""""""""""""
+
+The ``Series.flush()`` method for ``load_chunk`` and ``store_chunk`` as well as particle patch ``load``/``store`` is now only needed in ``defer``-red flush mode.
```
MINOR
I find it a bit misleading to limit the use of `Series.flush()` to dataset IO here.
This sentence is undoubtedly geared toward the user (rather than the dev), so what do you think about 

> The call to ``Series.flush()`` required for [...]

?

anokfireball commented on NEWS.rst:
```diff
@@ -3,6 +3,56 @@
+``.flush()`` Method
+"""""""""""""""""""
+
+The ``Series::flush()`` method for ``loadChunk`` and ``storeChunk`` as well as particle patch ``load``/``store`` is now only needed in ``DEFER``-red flush mode.
```
Same here.

anokfireball commented on examples/3_write_serial.py:
```diff
@@ -43,16 +46,12 @@
 
-    print("Stored the whole Dataset contents as a single chunk, " +
-          "ready to write content")
+    # in Flush_Type.defer, call now:
```
Maybe keep the old code and branch it depending on `FlushType`

```Python
if series.flush == openpmd_api.Flush_Type.defer:
    print("Stored the whole Dataset contents as a single chunk, ready to write content")
    # need to manually initiate transfer of accumulated operations
    series.flush()
    print("Dataset content has been fully written")
else:
    # transfer was implicitly done
    print("Stored the whole Dataset contents as a single chunk")
```

(similar branching code could be included in all comments concerning `.flush()` for `DEFER`)

ax3l replied:
The idea is to keep the code valid and compact. Branching here is not be best impression for copy-&-paste examples for my taste, since 100+ lines of code examples are hard to read. Just subjective feeling.
But one can do that, yep. Will think about it.

anokfireball commented on include/openPMD/IO/FlushType.hpp:
```diff
@@ -0,0 +1,42 @@
+    /** Data flush mode.
+     *
+     * Control if RecordComponent::load_chunk and RecordComponent::store_chunk
+     * operations are flushed immediately or delayed until Series::flush
```
Some of the `.flush()`es you removed in the documentation/examples do IO that is not concerned with chunk transfers:

```cpp
/* 3_write_serial.cpp:69 */

// in FlushType::DEFER, call now:
// series.flush();
cout << "File structure and required attributes have been written\n";
```

As I think this docstring is correct, some of the `print()`s after the commented-out code are now invalid.

anokfireball commented on include/openPMD/IO/FlushType.hpp:
```diff
@@ -0,0 +1,42 @@
+     * mode and ensure passed data buffers are valid and unmodified until
+     * Series::flush is called.
+     */
+    enum class FlushType //! @todo we should call this AccessMode and FlushMode ...
```
Another suggestion w.r.t. naming: `FlushStrategy`.

Generally, I support dropping `-Type`.

ax3l replied:
Uh, strategy is a looong word :D
If we rename, I would do so for both in a follow-up PR. Just dropping the suffix is also a great idea!

anokfireball commented on include/openPMD/IO/HDF5/HDF5Auxiliary.hpp:
```diff
@@ -253,7 +254,13 @@ concrete_h5_file_position(Writable* w)
+
+        auto afp_h5 = std::dynamic_pointer_cast< HDF5FilePosition >(afp);
+        if( afp_h5 == nullptr )
+            std::cerr << "WARNING: invalid file position in writable stack!" << std::endl;
```
Once we hit the `nullptr`, we're in an undetermined and likely unrecoverable state.
Rather than merely `cerr`ing and failing later down the line (i.e. when HDF5 fails to open a path), this makes a great candidate for `VERIFY`.

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -114,6 +116,14 @@ class RecordComponent : public BaseRecordComponent
      * record component will be selected.
+     *
+     * In FlushType::DEFER, the data managed by the returned
+     * std::shared_ptr< T > MUST NOT be read, modified or deleted until
```
Modification would probably not even be detrimental, but rather useless.

ax3l replied:
Yes, I basically just describe the contract here. Both is not allowed from a user-side for various reasons (either ignored or even crashes).

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -114,6 +116,14 @@ class RecordComponent : public BaseRecordComponent
+     *
+     * @todo implement: if the shared pointers' `use_count` drops to one
+     *       before Series::flush is called, the load will be skipped during
+     *       Series::flush.
```
Whoop! :raised_hands:  This is a great optimization.

ax3l replied:
Yes, but really needs #470 to work :)

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -137,9 +156,38 @@ class RecordComponent : public BaseRecordComponent
+     *
+     * @todo implement: if the shared pointers' `use_count` drops to one
+     *       before Series::flush is called, the store will be skipped during
+     *       Series::flush.
```
I don't think this works, though.
The user schedules a chunk store and then drops *his* ref to the data. So the write would still be valid.

ax3l replied:
Ah true for writes! So we can leave it here as is.

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -137,9 +156,38 @@ class RecordComponent : public BaseRecordComponent
 
+    /** Store a chunk of data
+     *
+     * shared_ptr< T > for data must be contiguous and large enough for extent
```
> and large enough for extent

This does not apply here or is misworded. It must have the same number of entries as the cumulative product of extent. *large enough* does not portray this fact.

**EDIT** Thinking more about it, you could only schedule part of a chunk in RAM. Still, I think *large enough* is not sufficient here.

ax3l replied:
ok, makes sense to be a more explicit.

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -137,9 +156,38 @@ class RecordComponent : public BaseRecordComponent
 
+    /** Store a chunk of data
+     *
+     * Passed contiguous container must be large enough for extent.
```
> and large enough for extent

See above.

anokfireball commented on include/openPMD/IO/FlushType.hpp:
```diff
@@ -0,0 +1,42 @@
+     */
+    enum class FlushType //! @todo we should call this AccessMode and FlushMode ...
+    {
+        DIRECT,  //!< immediately load/store data to passed data buffers (default)
```
Another suggestion w.r.t. naming: `IMMEDIATE`.

ax3l replied:
yes, I was looking for short, concise words so users have fun typing them.

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -275,6 +324,16 @@ RecordComponent::loadChunk(std::shared_ptr< T > data, Offset o, Extent e, double
+
+        if( this->IOHandler->flushType == FlushType::DIRECT )
+        {
+            while( !m_chunks->empty() )
```
The loop is probably redundant. The only scenario in which there would be more than one chunks in the `m_chunks` queue is when one chunk (or more) is scheduled in `DEFER`, the Series is switched to `DIRECT` and another chunk is scheduled.

ax3l replied:
Yes, that could happen imho and is what I had in mind.

anokfireball commented on include/openPMD/Series.hpp:
```diff
@@ -230,6 +231,25 @@ class Series : public Attributable
+     *
+     * Default: FlushType::DIRECT
+     */
+    void setFlush(FlushType flushType);
```
Nomenclature:
```cpp
void setFlushType(FlushType flushType);
```
?

ax3l replied:
and in the future
```C++
constructor(..., Access access, ...);
void setFlush(Flush flush);
Flush getFlush();
Access getAccess();
void flush();
```
?

anokfireball commented on include/openPMD/backend/PatchRecordComponent.hpp:
```diff
@@ -56,14 +57,37 @@ class PatchRecordComponent : public BaseRecordComponent
     //! @todo add support for constant patch record components
 
+    //! @todo allow to read all patch records at once in a convienient object
+    //!       to iterate over
```
There's an issue for this already (https://github.com/openPMD/openPMD-api/issues/361), maybe mention it here.

anokfireball commented on src/Series.cpp:
```diff
@@ -350,6 +350,16 @@ Series::setName(std::string const& n)
 
+void Series::setFlush(FlushType flushType)
+{
+    IOHandler->flushType = flushType;
```
To solve the redundant loop, we could `Series::flush()` when switching from `DEFER` TO `DIRECT`. This will schedule and perform all chunk transfers that would be done in the loops.

ax3l replied:
I think of the loop as a safety net at negligible costs.
Nevertheless, we can start flushing around here as well, that makes sense.

anokfireball commented on src/Series.cpp:
```diff
@@ -350,6 +350,16 @@ Series::setName(std::string const& n)
     return *this;
 }
 
+void Series::setFlush(FlushType flushType)
```
Maybe we want to include `FlushType` as an argument in the `Series::Series()` constructor with `DIRECT` as default to make the behaviour a little more obvious?

ax3l replied:
I thought about this and felt it has enough arguments.
If one wants to activate the "advanced" usage, I kinda like that it is an explicit step.

anokfireball reviewed (approved):
Thank you for improving user-approachability!

anokfireball commented on examples/3_write_serial.cpp:
```diff
@@ -63,15 +66,17 @@ int main(int argc, char *argv[])
 
-    series.flush();
+    // in FlushType::DEFER, call now:
+    // series.flush();
```
This commented out `flush()` is very likely what's causing the SEGFAULT.
In `RecordComponent::store_chunk()`, you only schedule a single `DATASET_WRITE` (well, multiple ones potentially).
Literally no other IO operations have been scheduled or completed at that point. No HDF5 handle exists, no structure inside the HDF5 file. So naturally the call in the backend fails ([here](https://github.com/openPMD/openPMD-api/blob/0.7.1-alpha/src/IO/HDF5/HDF5IOHandler.cpp#L670-L673)).

ax3l replied:
> This commented out `flush()` is very likely what's causing the SEGFAULT.

Yes, I know, that's why I was wondering if we want to [re-structure some of the logic](https://github.com/openPMD/openPMD-api/pull/484#issuecomment-466646641) in side `Record::flush_impl` so `RecordComponent` knows about it?

anokfireball reviewed (changes requested):
Your current changes do not entirely spare the user of calling `.flush()`.
As a very crude solution, you would call `Series.flush()` inside `store_chunk()` and `load_chunk()` (avoiding infinite recursion).

ax3l commented on NEWS.rst:
```diff
@@ -3,6 +3,56 @@
 Upgrade Guide
 =============
 
+0.8.0-beta
```
memo to myself: update me

ax3l commented on CHANGELOG.rst:
```diff
@@ -17,6 +17,7 @@ Changes to "0.7.1-alpha"
 Features
 """"""""
 
+- Series: flush type "direct" (new default) and "defer" (previous default) #484
```
move to newer release

ax3l commented:
@C0nsultant if you have the time, the logic in flushing here and in #490 needs to be refactored by us to be more flexible with flushes triggered from a record component (instead of top-down from a series-iteration-record-...) :-)
